subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/AbstractNamedFileSnapshotTaskStateChanges.java

```diff
@@ -55,6 +55,10 @@ protected AbstractNamedFileSnapshotTaskStateChanges(String taskName, TaskExecuti
         this.title = title;
         this.fileProperties = fileProperties;
         this.normalizationStrategy = normalizationStrategy;
+        retakeSnapshots();
+    }
+
+    protected void retakeSnapshots() {
         this.currentSnapshots = buildSnapshots(taskName, snapshotterRegistry, title, fileProperties);
```

Comment:
I think this has no effect if the `FileSystemMirror` is not flushed via `beforeTaskOutputsGenerated`. This happens if we enter [`LoadCommand.load`](https://github.com/gradle/gradle/blob/1e63343349a40180ad49926514589c20fed60984/subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TaskOutputCacheCommandFactory.java#L86-L86), which we do when we mess with the output files at all. Afterwards, we call `beforeTaskOutputsGenerated` again, just before executing the task [here](https://github.com/gradle/gradle/blob/0ef4638a07d3eefcee90bfbb730f164ed1ea566b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java#L66-L66). I guess retaking the snapshot has no other effect as deciding if a file is an output of a task or not. Since this shouldn't be a problem for cacheable tasks anyway, I think we can just not retake the snapshots here.
On the other hand, retaking the snapshots should be fast since all the output files would be empty/missing. I would also be interested in having an integration test which would fail if we do not retake the snapshots. Do we have such a test already?